### ISSUE TITLE
AttributeError:  'ConfigData' object has no attribute 'get' in Python 3.5.2 (Xenial)

### ISSUE THREAD

**Issue by chanchiwai-ray** [canonical/operator#827]
It appears that `typing.Mapping` works a bit different in Python version 3.5.2 than other higher versions. I got the following error when running some functional tests on xenial, but it works on bionic.

```python3
Traceback (most recent call last):
  File "./src/charm.py", line 310, in <module>
    main(CloudstatsCharm)
  File "/var/lib/juju/agents/unit-cloudstats-0/charm/venv/ops/main.py", line 431, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-cloudstats-0/charm/venv/ops/main.py", line 142, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-cloudstats-0/charm/venv/ops/framework.py", line 316, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-cloudstats-0/charm/venv/ops/framework.py", line 784, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-cloudstats-0/charm/venv/ops/framework.py", line 857, in _reemit
    custom_handler(event)
  File "./src/charm.py", line 145, in on_config_changed
    self.helper.configure()
  File "/var/lib/juju/agents/unit-cloudstats-0/charm/lib/lib_cloudstats.py", line 242, in configure
    self.render_config()
  File "/var/lib/juju/agents/unit-cloudstats-0/charm/lib/lib_cloudstats.py", line 118, in render_config
    context.update(self._openstack_context)
  File "/var/lib/juju/agents/unit-cloudstats-0/charm/lib/lib_cloudstats.py", line 146, in _openstack_context
    cert_config = self.charm_config.get("openstack_cert", None)
AttributeError: 'ConfigData' object has no attribute 'get'
```

Also see a simple code snippet

```python3
Python 3.5.2 (default, Jan 26 2021, 13:30:48) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing import Mapping
>>> class foo(Mapping[str, str]):
...     def __init__(self):
...             self.data = {"greeting": "Hello World"}
...     def __len__(self):
...             return len(self.data)
...     def __iter__(self):
...             return iter(self.data)
...     def __getitem__(self, key: str) -> str:
...             return self.data[key]
... 
>>> f = foo()
>>> f.get("greeting")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'foo' object has no attribute 'get'
>>> f["greeting"]
'Hello World'
```

**Comment by PietroPasotti** [comment]
Thanks for reporting this! 
I thought the recently merged smoke tests would catch this kind of bugs, but I guess we don't do full mapping api checks for config data. I guess this is one of the subtle ways a mapping is not a dict (and probably a change in the typing lib more than in python itself).

@rwcarlsen @pengale I'm not sure as what the best way to address this issue is. I couldn't find documentation clarifying what the difference is between collections.abc.Mapping between 3.5 and greater versions, so we can't be sure either that that's the _only_ difference.

According to the docs, the methods that Mapping provides are: `__contains__, keys, items, values, get, __eq__, __ne__`

I see three ways forward:
- drop Mapping inheritance, define 'manually' all the methods we want the object to have
- override `get` (perhaps conditionally on python version being < 3.6?) only and pray that's the only difference, or find a way to test it -- e.g. add a `assert set(dir(charm.config).keys()) == set(("get", "__getattr__", ...))`

Final considerations:
- probably StoredDict has the same issue, as it's using MutableMapping
- I thought collections.Mapping and typing.Mapping were basically aliases, but... not quite: their `dir()` differs:
```python
# dir(typing.Mapping()) = 
['__abstractmethods__', '__class__', '__class_getitem__', '__contains__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__module__', '__ne__', '__new__', '__orig_bases__', '__parameters__', '__reduce__', '__reduce_ex__', '__repr__', '__reversed__', '__setattr__', '__sizeof__', '__slots__', '__str__', '__subclasshook__', '__weakref__', '_abc_impl', '_is_protocol', 'data', 'get', 'items', 'keys', 'values']
# dir(collections.abc.Mapping()) = 
['__abstractmethods__', '__class__', '__contains__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__reversed__', '__setattr__', '__sizeof__', '__slots__', '__str__', '__subclasshook__', '__weakref__', '_abc_impl', 'data', 'get', 'items', 'keys', 'values']

**Comment by PietroPasotti** [comment]
So, after some digging, it looks like there are a bunch of differences in the Mapping `dir()` between 3.5.2 and greater versions.
Notably, all 'classic' dict public methods `get`, `items`, `keys`, `values` are absent in 3.5.2.

**Comment by benhoyt** [comment]
It's very strange how "they" thought it was a good idea to change this in a patch version (between 3.5.2 and 3.5.3)!

**Comment by rwcarlsen** [comment]
I'd vote for conditionally providing our own `.get` implementation for python<=3.5.2.

**Comment by PietroPasotti** [comment]
> I'd vote for conditionally providing our own `.get` implementation for python<=3.5.2.

Yep, but I already see three new issues coming up in the future:

AttributeError: 'ConfigData' object has no attribute 'items' in Python 3.5.2 (Xenial)
AttributeError: 'ConfigData' object has no attribute 'values' in Python 3.5.2 (Xenial)
AttributeError: 'ConfigData' object has no attribute 'keys' in Python 3.5.2 (Xenial)

So if we do that, we should also cover those. Or perhaps we can be a bit clever and do:
```python

# only on python < 3.5.2:
def __getattribute__(self, item):
    try:
         return super().__getattribute__(self, item)
    except AttributeError: 
        # give dict a chance
        return getattr(dict, item)
```

**Comment by PietroPasotti** [comment]
> 3.5.2

Somehow that release wasn't so minor: they added the typing module!
https://www.python.org/downloads/release/python-353/

**Comment by rwcarlsen** [comment]
> > 3.5.2
> 
> Somehow that release wasn't so minor: they added the typing module! https://www.python.org/downloads/release/python-353/

[facepalm]

**Comment by rwcarlsen** [comment]
Yeah - let's cover all four.

**Comment by benhoyt** [comment]
This was fixed in 1.5.3 for Xenial. In addition, we're no longer supported Python 3.5 in ops 2.0 and later (it's Python 3.8+ with Bionic).

**Comment by tonyandrewmeyer** [comment]
> Was that fixed in #810 (`1.5.2`)? It looks like it didn't include ConfigData as per the original report. With `1.5.5`, I'm still seeing these issues:

Indeed [1.5.5 LazyMapping](https://github.com/canonical/operator/blob/5bbc50faa49114cc6c4e0ae2b4b93351c33b6bbd/ops/model.py#L513) still relies on `Mapping` for `get`.

However, we don't support ops 1.x any more. As a workaround, you could do:

```python
try:
    self.env["http_proxy"] = self.config["override-http-proxy"]
except KeyError:
    self.env["http_proxy"] = self.env.get(...
```

**Comment by hloeung** [comment]
Okay, will use the workaround provided. Thanks @tonyandrewmeyer.
